### PR TITLE
v0.14.13 — ratings & reliability bug-fixes

### PR DESCRIPTION
**Pull request by Priapos1004**
## Summary

v0.14.13 is a bug-fix batch across ratings UX, search, and backend reliability. Rating forms gain not-yet-aired / no-episode-total guards and shared components (AttributeSelect / RatingNeighbors / ScoreDial); watch-time stats now credit actual episodes watched for every status. Title search is case-insensitive; the scraper waives popular short-form + feature-length ONAs. An `update_sweep` circuit breaker aborts on a total MAL outage instead of holding maintenance for hours, and a sliding-session idle-timeout keeps active users signed in while warning idle ones.

## Changes

- **Ratings**: shared `AttributeSelect` / `RatingNeighbors` / `ScoreDial`; actual-watch-time stats (`episodes_watched × duration_seconds`, every status); guards for not-yet-aired + no-episode-total media (single **and** bulk paths); rating-attribute scale enum migration (`has_3d_animation` rebucket, `ending_quality` neutral midpoint)
- **Search**: case-folded embeddings (capitalisation can't reorder title results); mobile-tappable search bar; scraper popularity + feature-length-ONA waivers; `full_story` recap → summary demotion
- **Reliability / UX**: `update_sweep` outage circuit breaker; admin Jobs Log live-refresh (+ pager-race fix); sliding-session idle-timeout + `/auth/refresh` + cross-tab token sync; job-bell completion toasts + clickable finished rows + dismissible badge; ratings-chart precision / clickable-activity / cumulative-to-now fixes; curation titles honour the name-language setting

## Quality Checks (ISO 25010)

- [x] Linting passed (`ruff` + `svelte-check`, both clean)
- [x] Tests passed (backend **728**, frontend **281**)
- [x] Functional Suitability — 6-agent review; two edge cases found, one fixed (bulk unaired guard), one deferred (see follow-up)
- [x] Performance Efficiency — no N+1 in the score-items projection; watch-time rewrite stays O(n)
- [x] Reliability — outage breaker verified end-to-end; Jobs Log pager race fixed
- [x] Security — `/auth/refresh` rejects expired tokens + re-reads role from DB; cross-tab sync is same-origin safe
- [x] Maintainability — dead-code / redundant-import checks clean; score-dial duplication resolved (`ScoreDial`)
- [x] Flexibility — attribute-scale migration reversible/safe; token-lifetime lower bound documented vs the frontend session constants
- [x] UI verified (hand-off — rating dial in RatingCard + BulkRateDialog; Jobs Log pager while a job runs)

## Known follow-up

- "Completed" is blocked for finished media with **NULL episodes** (`RatingCard.canComplete = totalEpisodes !== null`), and `episodesWatchedOf` credits such a completed rating **0** watch time (contradicting the completed→full-run fallback). Needs a design call (gate on airing status vs. episode presence). Recorded in the compound doc.

## Compound Doc

See: `compound-docs/2026-07-17-v0.14.13-bug-fixes.md`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: phsar/app/models/ratings.py [hunk 1]
```diff
@@ -57,9 +61,15 @@ class EndingType(str, enum.Enum):
     open = "open"
     closed = "closed"
     cliffhanger = "cliffhanger"
+    # Auto-set sentinel (never user-selectable) — set when the watch wasn't finished
+    # (on_hold/dropped), alongside ending_quality. v0.14.13.
+    not_applicable = "not_applicable"
 
 class EndingQuality(str, enum.Enum):
     unsatisfying = "unsatisfying"
+    # Middle anchor (v0.14.13) so an indifferent "it was okay" ending has its own value
```
Comment (Priapos1004) [inline phsar/app/models/ratings.py]: I see that it is important to keep track of changes, but should the comments be historic records as well, or only describe the current state? At least if the change is like here nothing major that one could accidently reverse

File: phsar/app/services/rating_service.py [hunk 2]
```diff
@@ -358,9 +387,18 @@ async def bulk_upsert_ratings(db: AsyncSession, user_id: int, data: RatingBulkCr
...
+    # so the note lands on the most recent season the user sees last in the table,
+    # regardless of the order media were clicked.
+    def _chrono_note_key(m: Media):
+        season = m.anime_season_name.value if m.anime_season_name else None
+        return chronological_media_key(m.anime_season_year, season, m.mal_id)
+
+    mains = [(i, m) for i, m in enumerate(media_list) if m.relation_type.value == "main"]
+    pool = mains or list(enumerate(media_list))
+    note_index = max(pool, key=lambda im: _chrono_note_key(im[1]))[0]
     shared_fields = data.model_dump(exclude=_EXCLUDE_BULK | {"note", "episodes_watched"})
```
[the remark this answers is not in the document]
Comment (Priapos1004) [reply phsar/app/services/rating_service.py]: Good catch on the inconsistency. Resolved by enforcing the bulk contract rather than honoring per-media input: bulk rating is a whole-anime "I finished this" action, so `watch_status` and `episodes_watched` are excluded from `shared_fields` and pinned per-media to `completed` / `media.episodes`. A direct/stale payload carrying `dropped` + a partial count is now normalized to completed + full-run instead of producing a "dropped but watched the full run" row — per-media watch state stays a single-rating concern (it makes no sense to bulk-mark several seasons of one anime as dropped). Documented the contract in the docstring and added a regression test (`test_bulk_upsert_normalizes_watch_state_to_completed_full_run`).

File: phsar/frontend/src/lib/utils/ratingStats.ts [hunk 3]
```diff
@@ -648,11 +656,43 @@ export function cumulativeWatchTime(items: RatingScoreItem[], sinceISO?: string)
...
+/** Estimated seconds the user actually watched of one media: the episodes they
+ * watched × the per-episode runtime (`duration_seconds`). Counted for every watch
+ * status, so a partly-watched on-hold/dropped series contributes its real time
+ * rather than zero. Returns 0 when we can't estimate (no per-episode runtime — e.g.
+ * a show with unknown duration), so the result is always finite and non-negative.
+ *
+ * episodes_watched is clamped to [0, episodes] when the catalog total is known, so a
+ * stale over-count can't exceed the full runtime; a completed row with no recorded
+ * episodes_watched (legacy, predating per-status counts) is assumed to have watched
+ * the full run. duration_seconds is used directly (not total_watch_time / episodes)
```
[the remark this answers is not in the document]
Comment (Priapos1004) [reply phsar/frontend/src/lib/utils/ratingStats.ts]: Aligned the docstring with the behavior: the completed→full-run fallback applies when the catalog episode total is known, else 0 (`episodes ?? 0`). No behavior change — `watchedSeconds` already returned 0 for the unknown-total case; this just makes the doc match.

File: phsar/frontend/src/lib/utils/ratingStats.ts [hunk 4]
```diff
@@ -648,11 +656,43 @@ export function cumulativeWatchTime(items: RatingScoreItem[], sinceISO?: string)
...
+	if (!dur || dur <= 0) return 0;
+	const eps = it.episodes;
+	let watched = episodesWatchedOf(it.episodes_watched, it.watch_status, eps);
+	if (eps != null && eps > 0) watched = Math.min(eps, watched); // clamp to known total
+	return watched * dur;
+}
+
+/** Episodes the user watched of one media, with the completed→full-run fallback for
+ * rows that predate per-status episode counts (a completed rating with no recorded
+ * count is assumed to have watched the full run). Always non-negative. Clamping to the
```
[the remark this answers is not in the document]
Comment (Priapos1004) [reply phsar/frontend/src/lib/utils/ratingStats.ts]: Aligned the docstring with the behavior: the completed→full-run fallback applies when the catalog episode total is known, else 0 (`episodes ?? 0`). No behavior change — `watchedSeconds` already returned 0 for the unknown-total case; this just makes the doc match.

File: phsar/app/services/scrape_dispatcher.py [hunk 5]
```diff
@@ -353,8 +359,28 @@ async def update_sweep_dispatcher(session: AsyncSession, job: Job) -> dict:
...
+                    if (
+                        consecutive_upstream_failures
+                        >= settings.JOBS_SWEEP_ABORT_AFTER_CONSECUTIVE_FAILURES
+                    ):
+                        raise TransientUpstreamError(
+                            f"update_sweep after {consecutive_upstream_failures} "
+                            "consecutive upstream failures — MAL appears to be "
+                            "unavailable"
+                        )
                 continue
```
[the remark this answers is not in the document]
Comment (Priapos1004) [reply phsar/app/services/scrape_dispatcher.py:445]: This is intentional (documented in the inline comment + services CLAUDE.md). The breaker exists to detect a *sustained* MAL outage; a stray non-upstream failure (e.g. a 404 on one bad row) doesn't prove MAL is usefully reachable, so only a genuine success resets the streak. Resetting on any non-upstream failure would make the breaker less likely to trip during an intermittent degradation that mixes 504s and occasional 404s — the opposite of what we want. Leaving as-is.

File: phsar/app/seeders/embedding_backfiller.py [hunk 6]
```diff
@@ -108,3 +119,75 @@ async def backfill_embeddings(db: AsyncSession):
...
+            db,
+            anime_id=anime.id,
+            title_texts=anime_title_texts(anime),
+            description_text=anime.description or "",
+        )
+        if i % _REEMBED_COMMIT_BATCH == 0:
+            await db.commit()
+            logger.info(f"Re-embed anime: {i}/{len(rows)}")
+    await db.commit()
+    return len(rows)
```
[the remark this answers is not in the document]
Comment (Priapos1004) [reply phsar/app/seeders/embedding_backfiller.py:137]: Declining this one: asyncpg buffers the full result set into memory regardless, so iterating the `ScalarResult` directly instead of `.all()` doesn't reduce peak memory — it only drops the Python list container, negligible next to the ORM objects the session tracks anyway. A genuine streaming fix would need `db.stream()` server-side cursors. We also use `len(rows)` for the `i/{len(rows)}` progress log and the return count, which a bare iterator loses. This path is a rare, env-gated (`EMBEDDING_REEMBED_ON_STARTUP`) one-time startup re-encode, so the trade-off isn't worth it.

File: phsar/app/seeders/embedding_backfiller.py [hunk 7]
```diff
@@ -108,3 +119,75 @@ async def backfill_embeddings(db: AsyncSession):
...
+            db,
+            media_id=media.id,
+            title_texts=media_title_texts(media),
+            description_text=media.description or "",
+        )
+        if i % _REEMBED_COMMIT_BATCH == 0:
+            await db.commit()
+            logger.info(f"Re-embed media: {i}/{len(rows)}")
+    await db.commit()
+    return len(rows)
```
[the remark this answers is not in the document]
Comment (Priapos1004) [reply phsar/app/seeders/embedding_backfiller.py:153]: Declining this one: asyncpg buffers the full result set into memory regardless, so iterating the `ScalarResult` directly instead of `.all()` doesn't reduce peak memory — it only drops the Python list container, negligible next to the ORM objects the session tracks anyway. A genuine streaming fix would need `db.stream()` server-side cursors. We also use `len(rows)` for the `i/{len(rows)}` progress log and the return count, which a bare iterator loses. This path is a rare, env-gated (`EMBEDDING_REEMBED_ON_STARTUP`) one-time startup re-encode, so the trade-off isn't worth it.

File: phsar/app/seeders/embedding_backfiller.py [hunk 8]
```diff
@@ -108,3 +119,75 @@ async def backfill_embeddings(db: AsyncSession):
...
+
+async def _reembed_ratings(db: AsyncSession) -> int:
+    rows = (await db.execute(select(Ratings).where(Ratings.note.isnot(None)))).scalars().all()
+    for i, rating in enumerate(rows, start=1):
+        await regenerate_rating_embedding(db, rating_id=rating.id, note=rating.note)
+        if i % _REEMBED_COMMIT_BATCH == 0:
+            await db.commit()
+            logger.info(f"Re-embed ratings: {i}/{len(rows)}")
+    await db.commit()
+    return len(rows)
```
[the remark this answers is not in the document]
Comment (Priapos1004) [reply phsar/app/seeders/embedding_backfiller.py:164]: Declining this one: asyncpg buffers the full result set into memory regardless, so iterating the `ScalarResult` directly instead of `.all()` doesn't reduce peak memory — it only drops the Python list container, negligible next to the ORM objects the session tracks anyway. A genuine streaming fix would need `db.stream()` server-side cursors. We also use `len(rows)` for the `i/{len(rows)}` progress log and the return count, which a bare iterator loses. This path is a rare, env-gated (`EMBEDDING_REEMBED_ON_STARTUP`) one-time startup re-encode, so the trade-off isn't worth it.

File: phsar/app/models/ratings.py (hunk 1 above)
Comment (Priapos1004) [reply phsar/app/models/ratings.py]: My take: comments should describe the *current state*, not carry changelog tags — git blame already records the "when", and a reader doesn't need to know a value was *added* in v0.14.13, only that `neutral` is the deliberate midpoint. The exception is where the prior behavior prevents a regression: a migration that renamed/removed enum values, or a "don't do X because Y" guard — those are landmines worth the historical note.

So I stripped the `(v0.14.13)` tags from the purely-descriptive comments (this line, the `not_applicable` sentinel just above, and the frontend comparison-badge header) and kept them where they guard against a regression: `ThreeDAnimation` (documents the `partial→medium` / `full→heavy` enum *rename* migration — the old values no longer exist) and the config idle-clock doc pointer (sits right next to the "keep above ~8 min" constraint).
